File: solutions/go/circular-buffer/2/circular_buffer.go
package circular

import "errors"

type Buffer struct{
    buf							[]byte
    lenth, capac, start, end	int
}

func NewBuffer(size int) *Buffer {
    return &Buffer{
        buf:	make([]byte, size),
        capac:	size,
    }
}

func (b *Buffer) ReadByte() (r byte, e error) {
    if b.lenth == 0 {
        e = errors.New("Read from empty buffer!")
        return
    }
    r, b.start, b.lenth = b.buf[b.start], (b.start+1) % b.capac, b.lenth-1
    return
}

func (b *Buffer) WriteByte(c byte) (e error) {
    if b.lenth == b.capac {
        e = errors.New("Write on full buffer!")
        return
    }
    b.buf[b.end], b.end, b.lenth = c, (b.end+1) % b.capac, b.lenth+1
    return
}

func (b *Buffer) Overwrite(c byte) {
    if b.lenth == b.capac {
        b.ReadByte()
    }
    b.WriteByte(c)
    return
}

func (b *Buffer) Reset() {
    b.start, b.end, b.lenth = 0, 0, 0
    return
}

File: solutions/kotlin/circular-buffer/2/src/main/kotlin/CircularBuffer.kt
import kotlin.collections.ArrayDeque

class EmptyBufferException: IllegalStateException("Empty Buffer")

class BufferFullException: IllegalStateException("Full Buffer")

class CircularBuffer<T>(val size: Int) {

    private val queue = ArrayDeque<T>(size)
    private var len = 0
    private var readFrom = 0
    private var writeTo = 0

    fun read(): T {
        if (isEmpty) {
            throw EmptyBufferException()
        }
        return queue
            .get(readFrom)
            .also {
                --len
                ++readFrom
                readFrom %= size
            }
    }
    
    fun write(value: T) {
        if (isFull) {
            throw BufferFullException()
        }
        if (queue.size <= writeTo) {
            queue.add(writeTo, value)
        } else {
            queue.set(writeTo, value)
        }
        ++len
        ++writeTo
        writeTo %= size
    }

    fun overwrite(value: T) {
        if (isFull) {
            read()
        }
        write(value)
    }
    
    fun clear() {
        len = 0
        readFrom = 0
        writeTo = 0
    }

    private val isEmpty
        get() = len == 0

    private val isFull
        get() = len == size

}